### PR TITLE
added new capabilities

### PR DESCRIPTION
**Pull request by Upellift99**
Thank you for the last merged ;)

The checkbox _Send approved booking email_ wasn't working for me.

Also, I have created specific capabilities for easier delegation, especially if you want to leave a role that could only approve or remove a booking, but without being able to create a new event.

### CODE REVIEW

**Comment by SiimKirjanen** [comment]
Hey again. Thank you for the PR.
I tested checkbox _Send approved booking email_ functionality and it was working for me but actually not sure why its working. Because in PHP string "0" and "1" both should evaluate to true. So I should receive the email even if _Approved booking receipt email_ is turned off.
Anyway it should be fixed by just modifying https://github.com/SiimKirjanen/seatreg/blob/master/php/SeatregSubmitBookings.php#L341 if statement. It should be `$this->_sendApprovedBookingEmail === '1'`
Can you test/modify your PR?

File: php/seatreg_functions.php [hunk 1]
```diff
@@ -3602,4 +3604,18 @@ function seatreg_create_payment_log() {
...
+    $role->add_cap('manage_events');
+    $role->add_cap('manage_bookings');
+}
+
+register_deactivation_hook(__FILE__, 'remove_plugin_capabilities');
+function remove_plugin_capabilities() {
+	$role = get_role('administrator');
+	$role->remove_cap('manage_events');
+	$role->remove_cap('manage_bookings');
+}
```
Comment (SiimKirjanen) [inline php/seatreg_functions.php:3621]: Can we prefix the capability names with `seatreg_`? Just to be sure we wont remove capabilities added by other plugins.

**Comment by SiimKirjanen** [comment]
When testing your code locally I don't see SeatReg submenu items anymore on my administrator account. Can it be problem with the new capability rules?

**Comment by SiimKirjanen** [comment]
I think those register_activation_hooks are not triggering.

**Comment by Upellift99** [comment]
Hey,
Thank you for your reply! and, thank you for your great plugin :star_struck:
Yes, the new capability rules need to be loaded once:

```add_action('admin_init', 'add_plugin_capabilities');
function add_plugin_capabilities() {
	$role = get_role('administrator');
	$role->add_cap('seatreg_manage_events');
	$role->add_cap('seatreg_manage_bookings');
}
```

I'm checking if `register_activation_hook` will be loaded when upgrading the plugin, and I will also rename the new functions with the prefix `seatreg_`

I'm still testing `$this->_sendApprovedBookingEmail === '1'` because it keeps sending the e-mail even if the box is unchecked.

**Comment by SiimKirjanen** [comment]
I think you need to relocate the new cap rules to https://github.com/SiimKirjanen/seatreg/blob/master/seatreg.php#L63-L75
Then it seems to work

**Comment by SiimKirjanen** [comment]
Okay so now my user (administrator) has `seatreg_manage_events` and `seatreg_manage_bookings `capabilities. How can I remove `seatreg_manage_events `capability from my user so it wont affect other administrator users?

**Comment by Upellift99** [comment]
Actually, the idea is to assign the new capability `seatreg_manage_bookings` to users with another role than the administrator role.
For example, if I want to allow subscribers to approve the bookings, I can use the plugin [User Role Editor](https://wordpress.org/plugins/user-role-editor/) to add this capability to this role:
image

Then, with a subscriber account, you only see the booking menu:
image

**Comment by SiimKirjanen** [comment]
There is still one problem. The `register_activation_hook `will not run when plugin is updated.
One solution to the problem would be to use `upgrader_process_complete `action hook but it would produce another problem.
I think we should just create a function that adds those capabilities and then just call it in inside https://github.com/SiimKirjanen/seatreg/blob/master/seatreg.php. Also would be nice to use https://developer.wordpress.org/reference/classes/wp_role/has_cap/ inside of the function. So only add capability if they are not added already.

**Comment by Upellift99** [comment]
I used the SEATREG_DB_VERSION to check if we need to deal with the capabilities ; so it's skipped if we are not upgrading.

File: seatreg.php [hunk 2]
```diff
@@ -86,4 +88,33 @@ function seatreg_plugin_deactivate() {
...
+}
+function seatreg_update_routine( $seatreg_db_stored_version ) {
+	if ( version_compare($seatreg_db_stored_version, '1.28', '>') ) {
+		seatreg_capabilities_add();
+	}
+}
+$seatreg_db_stored_version = get_option('SEATREG_DB_VERSION');
+if ($seatreg_db_stored_version !== SEATREG_DB_VERSION) {
+	seatreg_update_routine($seatreg_db_stored_version);
+}
```
Comment (SiimKirjanen) [inline seatreg.php:120]: Not sure about this logic.
The `get_option('SEATREG_DB_VERSION');` will always return false as the option `SEATREG_DB_VERSION `does not exist. There is a constant called `SEATREG_DB_VERSION` but its the version of the plugin database (dont think thats suitable here ).

The solution is to remove `seatreg_capabilities_add();` call from the `seatreg_plugin_activate()` and add it to https://github.com/SiimKirjanen/seatreg/blob/master/seatreg.php#L47-L50 just before `require( 'php/seatreg_admin_panel.php' );	`

**Comment by Upellift99** [comment]
Thanks for your reply!
Yes, I used the wrong name of the variable.

Actually, `seatreg_capabilities_add()` is executed either when the plugin is activated, and either when the plugin is upgrading from now, and that's it! I'm not sure it's a good practice to execute `seatreg_capabilities_add()` everytime `is_admin()` is called. That was the idea.

File: seatreg.php [hunk 3]
```diff
@@ -86,4 +88,33 @@ function seatreg_plugin_deactivate() {
...
+function seatreg_update_routine( $seatreg_db_stored_version ) {
+	if ( version_compare($seatreg_db_stored_version, '1.28', '>') ) {
+		seatreg_capabilities_add();
+	}
+}
+$seatreg_db_stored_version = get_option('seatreg_db_current_version');
+if ($seatreg_db_stored_version !== SEATREG_DB_VERSION) {
+	seatreg_update_routine($seatreg_db_stored_version);
+}
```
Comment (SiimKirjanen) [inline seatreg.php:120]: This logic wont detect when plugin is upgrading. It detects when SeatReg database has changed.

**Comment by SiimKirjanen** [comment]
>  I'm not sure it's a good practice to execute `seatreg_capabilities_add()` everytime `is_admin()` is called. That was the idea.

Yes its not a good idea but its best solution I game up with. I just want to avoid any issues in the future. For example someone might want to remove those capabilities from the admins. Currently its possible but then next time they will be added back?

**Comment by Upellift99** [comment]
The administrator role in WordPress has all the capabilities; an admin would even be able to get back a capability that has been disabled for the administrator role :smile: 

Introducing these new capabilities are useful for all the other roles: the admin can choose which role will be able to create an event and/or validate booking (the Editor role, the Author role, the Subscriber role, or even any other custom role).

At this point, the only role which is able to use SeatReg is the administrator role. The new capabilities would help to delegate the use of SeatReg to other roles. In the future, upgrading the plugin won't have any effect on any role customization: we don't tell the plugin to remove capabilities on any role.

And the `is_admin()` function doesn't check if you are an admin, but if you are on the interface admin; so it might concern any role https://developer.wordpress.org/reference/functions/is_admin/

**Comment by SiimKirjanen** [comment]
I think its good idea to split this pull request to two. So capability change and the approved booking email fix.

**Comment by SiimKirjanen** [comment]
Lets remove the db change logic and try

```
$seatreg_trigger_side_effect = get_option('seatreg_trigger_side_effect');
if ($seatreg_trigger_side_effect !== SEATREG_TRIGGER_SIDE_EFFECT) {
	seatreg_capabilities_add();
	update_option( "seatreg_trigger_side_effect", SEATREG_TRIGGER_SIDE_EFFECT );
}
```
And in constants.php add

define('SEATREG_TRIGGER_SIDE_EFFECT ', '1');

**Comment by SiimKirjanen** [comment]
There are some merge conflicts

File: seatreg.php [hunk 4]
```diff
@@ -86,4 +88,29 @@ function seatreg_plugin_deactivate() {
...
+function seatreg_capabilities_remove() {
+	$role = get_role('administrator');
+	if ( $role->has_cap('seatreg_manage_events') ) {
+		$role->remove_cap('seatreg_manage_events');
+	}
+	if ( $role->has_cap('seatreg_manage_bookings') ) {
+		$role->remove_cap('seatreg_manage_bookings');
+	}
+}
```
Comment (SiimKirjanen) [inline seatreg.php:111]: What would you say if we relocate those functions to seatreg_functions.php file?

File: php/seatreg_admin_panel.php [hunk 5]
```diff
@@ -36,6 +36,7 @@ function seatreg_add_plugin_menu() {
 		/* translators: %s: Name of the registration */
 		sprintf(esc_html__('%s Overview', 'seatreg'), 'SeatReg'),  //page title
 		esc_html__('Overview', 'seatreg'),  //menu title
+		'seatreg_manage_events',  //capability
 		'manage_options',  //capability
```
Comment (SiimKirjanen) [inline php/seatreg_admin_panel.php:40]: Accidentally merged the old capability. Can you remove it? Not sure if I can.

**Comment by Upellift99** [comment]
Thank you for all the work you've made!

**Pull request merged**